sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreCheckpointFormatV2Suite.scala

```diff
@@ -203,14 +210,299 @@ class RocksDBStateStoreCheckpointFormatV2Suite extends StreamTest
+          StopStream
+        )
+
+        testStream(aggregated, Update)(
+          StartStream(checkpointLocation = checkpointDir.getAbsolutePath),
+          AddData(inputData, 3, 2, 1),
+          CheckLastBatch((3, 3), (2, 2), (1, 1)),
+          // By default we run in new tuple mode.
+          AddData(inputData, 4, 4, 4, 4),
+          CheckLastBatch((4, 4)),
+          AddData(inputData, 5, 5),
+          CheckLastBatch((5, 2)),
+          StopStream
+        )
+      }
+
+      withSQLConf(changelogEnabled, ckptv2) {
+        testStream(aggregated, Update)(
+          StartStream(checkpointLocation = checkpointDir.getAbsolutePath),
+          AddData(inputData, 4),
+          CheckLastBatch((4, 5)),
+          AddData(inputData, 4),
+          CheckLastBatch((4, 6)),
+          StopStream
+        )
+
+        testStream(aggregated, Update)(
+          StartStream(checkpointLocation = checkpointDir.getAbsolutePath),
+          AddData(inputData, 6, 7, 8),
+          CheckLastBatch((6, 1), (7, 1), (8, 1)),
+          AddData(inputData, 4, 4, 4, 4),
+          CheckLastBatch((4, 10)),
+          AddData(inputData, 5, 5),
+          CheckLastBatch((5, 4)),
+          StopStream
+        )
+      }
+
+      withSQLConf(changelogDisabled, ckptv2) {
+        testStream(aggregated, Update)(
+          StartStream(checkpointLocation = checkpointDir.getAbsolutePath),
+          AddData(inputData, 3, 2, 1),
+          CheckLastBatch((3, 4), (2, 3), (1, 2)),
+          AddData(inputData, 4),
+          CheckLastBatch((4, 11)),
+          StopStream
+        )
+      }
+    }
+  }
+
+  testConfigSetups.foreach { testConfig =>
+    testWithRocksDBStateStore("checkpointFormatVersion2 Compatibility - " +
+      s"transformWithState - $testConfig") {
+        withTempDir { checkpointDir =>
+          val inputData = MemoryStream[String]
+          val result = inputData.toDS()
+            .groupByKey(x => x)
+            .transformWithState(new RunningCountStatefulProcessor(),
+              TimeMode.None(),
+              OutputMode.Update())
+
+          withSQLConf(testConfig.configsForFirstRun: _*) {
+            testStream(result, Update())(
+              StartStream(checkpointLocation = checkpointDir.getAbsolutePath),
+              AddData(inputData, "a"),
+              CheckNewAnswer(("a", "1")),
+              Execute { q =>
+                assert(q.lastProgress.stateOperators(0)
+                  .customMetrics.get("numValueStateVars") > 0)
+                assert(q.lastProgress.stateOperators(0)
+                  .customMetrics.get("numRegisteredTimers") == 0)
+              },
+              AddData(inputData, "a", "b"),
+              CheckNewAnswer(("a", "2"), ("b", "1")),
+              StopStream
+            )
+            testStream(result, Update())(
+              StartStream(checkpointLocation = checkpointDir.getAbsolutePath),
+              // should remove state for "a" and not return anything for a
```

Comment:
curious, why does that happen?
oh... i see, `RunningCountStatefulProcessor` drops it when count gets to 3